File: Wells/Source/VievModels/WellsViewModel.vb
Imports System.Reflection
Imports Wells.YPFModel
Imports Wells.ViewBase
Imports NPOI.XSSF.UserModel

Public Class WellsViewModel
    Inherits EntitiesViewModel(Of Well)

    Private _Entities As IQueryable(Of Well)
    Private _SelectedEntity As Well
    Private _SelectedEntities As IEnumerable(Of Well)

    Sub New()
        MyBase.New(Nothing)
        IsNewCommandEnabled = True
        IsRemoveCommandEnabled = True
        FilterCollection = New FilterCollection(Of Well)
        Initialize()
        _Entities = _Repository.Wells.All
        _ShowWellPanel = True
    End Sub

    Public Overrides ReadOnly Property Entities As IEnumerable(Of Well)
        Get
            If FilterCollection?.Any Then
                Dim l = FilterCollection.Apply(_Entities)
                _EntitiesCount = l.Count
                NotifyEntityCount()
                Return l.ToList
            End If
            _EntitiesCount = _Entities.Count
            NotifyEntityCount()
            Return _Entities.ToList
        End Get
    End Property

    Public Overrides ReadOnly Property FilterProperties As Dictionary(Of String, PropertyInfo)
        Get
            Return Well.Properties
        End Get
    End Property

    Public Overrides Property SelectedEntity As Well
        Get
            Return _SelectedEntity
        End Get
        Set
            SetValue(_SelectedEntity, Value)
            NotifyPropertyChanged(NameOf(WellExistsInfo))
        End Set
    End Property

    Public Overrides Property SelectedEntities As IEnumerable(Of Well)
        Get
            Return _SelectedEntities
        End Get
        Set
            SetValue(_SelectedEntities, Value)
        End Set
    End Property

    Public Overrides ReadOnly Property EditEntityCommand As ICommand = New RelayCommand(Sub()
                                                                                            Dim vm As New EditWellViewModel(SelectedEntity)
                                                                                            If _Control.MainWindow.OpenEditEntityDialog(vm) Then
                                                                                                UpdateEntites()
                                                                                            End If
                                                                                        End Sub, Function() SelectedEntity IsNot Nothing, AddressOf OnError)

    Public Overrides ReadOnly Property IsNewCommandEnabled As Boolean

    Public Overrides ReadOnly Property IsRemoveCommandEnabled As Boolean

    Public Overrides ReadOnly Property NewEntityCommand As ICommand = New RelayCommand(Sub()
                                                                                           Dim vm As New EditWellViewModel()
                                                                                           If _Control.MainWindow.OpenEditEntityDialog(vm) Then
                                                                                               UpdateEntites()
                                                                                           End If
                                                                                       End Sub, Function() IsNewCommandEnabled, AddressOf OnError)

    Public Overrides ReadOnly Property ImportEntitiesCommand As ICommand = New RelayCommand(Sub()
                                                                                                Dim wb As XSSFWorkbook = Nothing
                                                                                                Dim sheetIndex As Integer = -1

                                                                                                If OpenExcelFile(wb, sheetIndex) Then
                                                                                                    ReadExcelFile(wb, sheetIndex)
                                                                                                    UpdateEntites()

                                                                                                End If
                                                                                            End Sub, Function() IsNewCommandEnabled, AddressOf OnError, AddressOf CloseWaitingMessage)

    Public Overrides ReadOnly Property RemoveEntityCommand As ICommand = New RelayCommand(Sub()
                                                                                              If _Control.MainWindow.ShowMessageBox("¿Está seguro de eliminar este pozo?", "Eliminar") Then
                                                                                                  _Repository.Wells.Remove(SelectedEntity)
                                                                                                  UpdateEntites()
                                                                                              End If
                                                                                          End Sub, Function() SelectedEntity IsNot Nothing AndAlso IsRemoveCommandEnabled, AddressOf OnError)

    Private Sub UpdateEntites()
        _Entities = _Repository.Wells.All
        NotifyPropertyChanged(NameOf(Entities))
    End Sub

    Private Async Sub ReadExcelFile(workbook As XSSFWorkbook, sheetIndex As Integer)
        ShowWaitingMessage("Leyendo pozos del archivo Excel...")
        Dim wells = Await Task.Run(Function() ExcelReader.ReadWells(workbook, sheetIndex, _progress))
        CloseWaitingMessage()

        If wells.Any Then
            ShowWaitingMessage("Importando pozos...")
            Await Task.Run(Sub() _Repository.Wells.AddRangeAsync(wells))
            CloseWaitingMessage()

            ShowWaitingMessage("Guardando base de datos...")
            Await _Repository.SaveChangesAsync()

        End If

        workbook.Close()
        CloseWaitingMessage()
    End Sub

    Protected Overrides Sub CreateWellFilter()
        Dim wellFilter = New WellFilter(Of Well)(_Repository.Wells, True, WellType, WellProperty, SelectedWellName)
        OnCreatingFilter(wellFilter)
    End Sub

    Public Overrides Function GetContextMenu() As ContextMenu
        Dim menu = New ContextMenu()
        Dim editMenuItem As New MenuItem() With {.Header = "Editar...", .Command = EditEntityCommand}
        Dim removeMenuItem As New MenuItem() With {.Header = "Eliminar", .Command = RemoveEntityCommand}
        menu.Items.Add(editMenuItem)
        menu.Items.Add(New Separator)
        menu.Items.Add(removeMenuItem)
        Return menu
    End Function

    Overrides ReadOnly Property WellExistsInfo As String
        Get
            If _SelectedEntity IsNot Nothing Then
                Return If(_SelectedEntity.Exists, "Pozo existente", "Pozo inexistente")
            End If
            Return String.Empty
        End Get
    End Property
End Class

File: Wells/src/ViewModels/Entities/WellsViewModel.cs
using Microsoft.VisualBasic;
using NPOI.XSSF.UserModel;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using System.Windows.Controls;
using System.Windows.Input;
using Wells.BaseView;
using Wells.BaseView.ViewInterfaces;
using Wells.Model;
using Wells.Persistence.Repositories;
using Wells.View.Filters;
using Wells.View.Graphics;
using Wells.View.Importer;

namespace Wells.View.ViewModels
{
    public class WellsViewModel : EntitiesViewModel<Well>
    {
        private List<PremadeSeriesInfoCollection> _PremadeGraphics;
        private readonly MenuItem filesMenuItem;

        public WellsViewModel() : base(null)
        {
            IsNewCommandEnabled = true;
            IsRemoveCommandEnabled = true;
            //_FilterCollection = new FilterCollection<Well>();
            ReadFilters(Information.TypeName(this));
            Initialize();
            _Entities = Repository.Wells.All;
            _ShowWellPanel = true;
            _PremadeGraphics = CreatePremadeGraphicViewModel.ReadPremadeGraphics();
            filesMenuItem = new MenuItem { Header = "Archivos" };
        }

        protected override void OnSetView(IView view)
        {
            base.OnSetView(view);
            MainWindow.PremadeGraphicsChanged += OnPremadeGraphicsChanged;
        }

        public override string WellExistsInfo
        {
            get
            {
                if (_SelectedEntity != null)
                {
                    return _SelectedEntity.Exists ? "Pozo existente" : "Pozo inexistente";
                }
                return string.Empty;
            }
        }

        public override bool IsNewCommandEnabled { get; }

        public override bool IsRemoveCommandEnabled { get; }

        public override ICommand NewEntityCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var vm = new EditWellViewModel();
                    if (MainWindow.OpenEditEntityDialog(vm))
                    {
                        UpdateEntites();
                    }
                }, (obj) => IsNewCommandEnabled, OnError);
            }
        }

        public override ICommand EditEntityCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var vm = new EditWellViewModel(SelectedEntity);
                    if (MainWindow.OpenEditEntityDialog(vm))
                    {
                        UpdateEntites();
                    }
                }, (obj) => SelectedEntity != null, OnError);
            }
        }

        public override ICommand RemoveEntityCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if (SharedBaseView.ShowYesNoMessageBox(MainWindow, "¿Está seguro de eliminar este pozo?", "Eliminar"))
                    {
                        Repository.Wells.Remove(SelectedEntity);
                        RepositoryWrapper.Instance.SaveChanges();
                        UpdateEntites();
                    }
                }, (obj) => SelectedEntity != null && IsRemoveCommandEnabled, OnError);
            }
        }

        public ICommand OpenPremadeGraphicCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    MainWindow.OpenGraphicsView(SelectedEntity, param as PremadeSeriesInfoCollection);
                }, (obj) => SelectedEntity != null, OnError);
            }
        }

        public ICommand OpenPiperShoellerGraphicCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if (SelectedEntities != null && SelectedEntities.Any())
                    {
                        var vm = new PiperSchoellerGraphicViewModel(SelectedEntities);
                        MainWindow.OpenGraphicsView(vm);
                    }
                    else if (SelectedEntity != null)
                    {
                        var vm = new PiperSchoellerGraphicViewModel(new List<Well> { SelectedEntity });
                        MainWindow.OpenGraphicsView(vm);
                    }
                }, (obj) => (SelectedEntities != null && SelectedEntities.Any()) || SelectedEntity != null, OnError);
            }
        }


        public ICommand OpenMapCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if (SelectedEntities != null && SelectedEntities.Any())
                    {
                        var vm = new MapViewModel(SelectedEntities);
                        MainWindow.OpenGraphicsView(vm);
                    }
                }, (obj) => SelectedEntities != null && SelectedEntities.Any(), OnError);
            }
        }

        public ICommand OpenFileCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if (param != null)
                    {
                        var file = param as ExternalFile;
                        file.Open();
                    }
                }, (obj) => true, OnError);
            }
        }

        public override Dictionary<string, PropertyInfo> FilterProperties => Well.Properties;

        public override Well SelectedEntity 
        { 
            get => _SelectedEntity; 
            set 
            { 
                SetValue(ref _SelectedEntity, value);
                UpdateFilesContextMenu();
                NotifyPropertyChanged(nameof(WellExistsInfo)); 
            } 
        }

        protected override void SetCommandUpdates()
        {
            base.SetCommandUpdates();
            Add(nameof(SelectedEntity), OpenPremadeGraphicCommand);
            Add(nameof(SelectedEntity), OpenPiperShoellerGraphicCommand);
            Add(nameof(SelectedEntities), OpenPiperShoellerGraphicCommand);
        }

        public override ContextMenu GetContextMenu()
        {
            var menu = new ContextMenu();
            var editMenuItem = new MenuItem { Header = "Editar...", Command = EditEntityCommand };
            menu.Items.Add(editMenuItem);

            menu.Items.Add(new Separator());
            menu.Items.Add(filesMenuItem);

            menu.Items.Add(new Separator());
            var exportMenuItem = new MenuItem { Header = "Exportar...", Command = ExportEntitiesCommand };
            menu.Items.Add(exportMenuItem);
            
            menu.Items.Add(new Separator());
            var graphicsMenuItem = new MenuItem { Header = "Gráficos" };
            var mapMenuItem = new MenuItem { Header = "Mostrar en mapa...", Command = OpenMapCommand, CommandParameter = SelectedEntities };
            graphicsMenuItem.Items.Add(mapMenuItem);
            graphicsMenuItem.Items.Add(new Separator());

            var piperMenuItem = new MenuItem { Header = "Piper-Schöeller", Command = OpenPiperShoellerGraphicCommand, CommandParameter = SelectedEntities };
            graphicsMenuItem.Items.Add(piperMenuItem);
            foreach (var pg in _PremadeGraphics)
            {
                graphicsMenuItem.Items.Add(new Separator());
                var aMenuItem = new MenuItem { Header = pg.Title, Command = OpenPremadeGraphicCommand, CommandParameter = pg };
                graphicsMenuItem.Items.Add(aMenuItem);
            }
            menu.Items.Add(graphicsMenuItem);

            menu.Items.Add(new Separator());
            var removeMenuItem = new MenuItem { Header = "Eliminar", Command = RemoveEntityCommand };
            menu.Items.Add(removeMenuItem);
            return menu;
        }

        private void UpdateFilesContextMenu()
        {
            filesMenuItem.Items.Clear();
            foreach(var file in SelectedEntity?.Files)
            {
                var aMenuItem = new MenuItem { Header = file.CompleteFilename, Command = OpenFileCommand, CommandParameter = file };
                filesMenuItem.Items.Add(aMenuItem);
            }
        }

        protected override async Task ReadExcelFile(XSSFWorkbook workbook, int sheetIndex)
        {
            ShowWaitingMessage("Leyendo pozos del archivo Excel...");
            var wells = await Task.Run(() => ExcelReader.ReadWells(workbook, sheetIndex, null));
            CloseWaitingMessage();

            if (wells.Any())
            {
                ShowWaitingMessage("Importando pozos...");
                await Task.Run(() => Repository.Wells.AddRangeAsync(wells));
                CloseWaitingMessage();

                ShowWaitingMessage("Guardando base de datos...");
                await Repository.SaveChangesAsync();
            }

            workbook.Close();
            CloseWaitingMessage();
        }


        void OnPremadeGraphicsChanged(object sender, EventArgs args)
        {
            _PremadeGraphics = CreatePremadeGraphicViewModel.ReadPremadeGraphics();
            Control.UpdateRowContextMenu();
        }

        protected override void CreateWellFilter()
        {
            var wellFilter = new WellFilter<Well>(true, WellType, WellProperty, SelectedWellName);
            OnCreatingFilter(wellFilter);
        }

        protected override void UpdateEntites()
        {
            _Entities = Repository.Wells.All;
            NotifyPropertyChanged(nameof(Entities));
        }
    }
}
